Add director

uncle-sly opened the pull request:
Добавление режиссёров в фильмы.

u-sly committed:
Задача: Добавление режиссёров в фильмы.
Добавлена сущность Director и для нее слои контролллера, сервиса и репозитория DirectorController, DirectorService, DirectorServiceImpl, DirectorRepository, JdbcDirectorRepository, а так же маппе DirectorRowMapper.

Film добавлено поле для Режиссеров.
FilmController добавлен метод для обработки endpoint.
FilmRepository, FilmService добавлено описание метода getSortedDirectorsFilms.
FilmExtractor добавлен маппинг режиссеров
FilmServiceImpl добавлена обработка логики для etSortedDirectorsFilms и обработка режиссеров в методе update.

JdbcBaseRepository изменен метод batchInsert , чтобы принимал сущности разных типов.

Для таблицы связей фильмов и режиссеров добавлены JdbcFilmDirectorRepository и FilmDirectorRowMapper.

JdbcFilmGenreRepository добавлены параметры в метод save

JdbcFilmRepository для обработки режиссеров
добавлена record для хранения сущности Film_Director, изменены метода get, getAll, update, добавлен метод getSortedDirectorsFilms

SortOrder добавлен enum для обработки способа сортировки фильмов режиссера.
schema.sql - добавлено таблица directors, film_director

data.sql - добавлена очистка и сброс счетчика для directors

u-sly committed:
Задача: Добавление режиссёров в фильмы.

рефакторинг FilmServiceImpl дублирующий код вынесен в метод fillUpGenres.
рефакторинг JdbcFilmRepository дублирующий код вынесен в метод fillUpDirectors.

Все тесты Postman локально прошли успешно.

u-sly committed:
Задача: Добавление режиссёров в фильмы.

FilmRepositoryTests
исправлен импорт бинов, скорретирован один тест.

u-sly committed:
Задача: Добавление режиссёров в фильмы.

ErrorResponse
заменено message на error.

Afaiwjs6272 commented on src/main/java/ru/yandex/practicum/filmorate/controller/DirectorController.java:
```diff
@@ -0,0 +1,59 @@
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.validation.annotation.Validated;
+import org.springframework.web.bind.annotation.*;
```
В инете читал, что лучше конкретно указывать какие анотации ты используешь

uncle-sly replied:
поправил

Afaiwjs6272 commented on src/main/java/ru/yandex/practicum/filmorate/repository/inDatabase/JdbcDirectorRepository.java:
```diff
@@ -0,0 +1,70 @@
+    public List<Director> getAll() {
+        String sql = "SELECT * FROM directors";
+
+        return findMany(sql, Map.of());
```
Можно ведь не создавать пустую мапу в качестве второго аргумента

uncle-sly replied:
к сожалению не нашел, чем заменить.. метод findMany ожидает мапу , но в данном случае параметров нет, поэтому передал пустую.. если есть идея чем заменить, подскажи пожалуйста.

Afaiwjs6272 commented on src/main/java/ru/yandex/practicum/filmorate/repository/inDatabase/JdbcDirectorRepository.java:
```diff
@@ -0,0 +1,70 @@
+        return findOne(sql, params);
+    }
+
+    public List<Director> getByIds(List<Long> ids) {
+        String list = ids.stream().map(String::valueOf)
+                .collect(Collectors.joining(","));
+        String sql = "SELECT * FROM directors WHERE id in (" + list + ")";
+
+        return findMany(sql, Collections.emptyMap());
+    }
```
Для удобства и красоты кода можно изменить вызов каждого метода в стриме с новой строки 
Например: .stream()
                   .map()
                   .collect()

uncle-sly replied:
поправил

Afaiwjs6272 commented on src/main/java/ru/yandex/practicum/filmorate/repository/inDatabase/JdbcFilmDirectorRepository.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.ArrayList;
+import java.util.Map;
+
+@Repository
+@Slf4j
+public class JdbcFilmDirectorRepository extends JdbcBaseRepository<Object> {
```
Мне кажется, что для читабельности кода нужно поменять анотации местами, чтобы было понято что конретный класс является репозиторием

uncle-sly replied:
согласен, поправил везде

Afaiwjs6272 commented on src/main/java/ru/yandex/practicum/filmorate/service/DirectorServiceImpl.java:
```diff
@@ -0,0 +1,51 @@
+
+import java.util.List;
+
+@Service
+@RequiredArgsConstructor
```
Тут я бы тоже поменял местами

uncle-sly replied:
поправил тоже

Afaiwjs6272 commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmServiceImpl.java:
```diff
@@ -85,6 +95,22 @@ public void addLike(long filmId, long userId) {
+
+        String order = SortOrder.from(sortBy);
+
+        if ("null".equals(order)) {
+            throw new ValidationException("Получено: " + sortBy + " Должно быть year или likes");
+        }
```
Я думаю, что для чистоты и понятности кода нужно конструкцию условие в if записать так - (orders.equals(null)), чтобы было понято что мы поверяем условие для order

uncle-sly replied:
поправил

Afaiwjs6272 commented on src/main/java/ru/yandex/practicum/filmorate/service/SortOrder.java:
```diff
@@ -0,0 +1,17 @@
+package ru.yandex.practicum.filmorate.service;
+
+public enum SortOrder {
+
+    year,
+    likes;
```
Насколько я знаю поля в enum мы записываем в верхнем регистре

uncle-sly replied:
да, спасибо, поправил

IvanSimonovDev commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmServiceImpl.java:
```diff
@@ -22,50 +20,41 @@ public class FilmServiceImpl implements FilmService {
-                throw new FkConstraintViolationException("Жанр вне диапазона.");
-            }
-        }
+        fillUpGenres(film);
```
Валидация жанров при создании фильма есть, а режиссёров нет

uncle-sly replied:
Спасибо, поправил

u-sly committed:
Добавление режиссёров в фильмы.

Поправлен импорт и порядок аннотаций:
DirectorController
DirectorServiceImpl
FilmController
GenreService
JdbcDirectorRepository
JdbcFilmDirectorRepository
JdbcFilmGenreRepository
JdbcFilmRepository
JdbcGenreRepository
JdbcMpaRepository
JdbcUserFriendRepository
JdbcUserRepository
MpaService
UserController
UserServiceImpl

FilmServiceImpl- дублирующий код вынесен в метод fillUpDirectors, поправлен порядок аннотаций и условие с order в getSortedDirectorsFilms

SortOrder - поправлены константы enum
